File: agents/schemas/dashboard_state.py
"""
Dashboard Shared State Schemas

These Pydantic models mirror the TypeScript schemas for state
shared between the Dashboard Gateway agent and the frontend.
The schemas use Field aliases for camelCase output to ensure
cross-language compatibility with the TypeScript frontend.

@see docs/modules/bm-dm/epics/epic-dm-04-tech-spec.md
Epic: DM-04 | Story: DM-04.1
"""

from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional, Union

from pydantic import BaseModel, ConfigDict, Field

# =============================================================================
# STATE VERSION
# =============================================================================

# State schema version for migrations.
# Must match the TypeScript STATE_VERSION constant.
STATE_VERSION = 1


# =============================================================================
# ENUMS
# =============================================================================


class ProjectStatus(str, Enum):
    """Project status values matching TypeScript ProjectStatusEnum."""

    ON_TRACK = "on-track"
    AT_RISK = "at-risk"
    BEHIND = "behind"
    COMPLETED = "completed"


class TrendDirection(str, Enum):
    """Metric trend direction matching TypeScript TrendDirectionEnum."""

    UP = "up"
    DOWN = "down"
    NEUTRAL = "neutral"


class AlertType(str, Enum):
    """Alert severity type matching TypeScript AlertTypeEnum."""

    ERROR = "error"
    WARNING = "warning"
    INFO = "info"
    SUCCESS = "success"


# =============================================================================
# TASK PROGRESS ENUMS AND MODELS (DM-05.4)
# =============================================================================


class TaskStepStatus(str, Enum):
    """Step execution status matching TypeScript TaskStepStatusEnum."""

    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class TaskStatus(str, Enum):
    """Overall task status matching TypeScript TaskStatusEnum."""

    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class TaskStep(BaseModel):
    """
    Individual step within a task.

    Tracks execution state and progress of a single step in a multi-step task.
    """

    model_config = ConfigDict(populate_by_name=True, use_enum_values=True)

    index: int = Field(..., ge=0, description="Step index (0-based)")
    name: str = Field(..., description="Step display name")
    status: TaskStepStatus = Field(
        default=TaskStepStatus.PENDING, description="Step execution status"
    )
    started_at: Optional[int] = Field(
        None, alias="startedAt", description="Step start timestamp (Unix ms)"
    )
    completed_at: Optional[int] = Field(
        None, alias="completedAt", description="Step completion timestamp (Unix ms)"
    )
    progress: Optional[int] = Field(
        None, ge=0, le=100, description="Sub-step progress percentage (0-100)"
    )


class TaskProgress(BaseModel):
    """
    Progress state for a long-running task.

    Tracks overall task status and individual step progress for
    real-time streaming to the frontend.
    """

    model_config = ConfigDict(populate_by_name=True, use_enum_values=True)

    task_id: str = Field(..., alias="taskId", description="Unique task identifier")
    task_name: str = Field(..., alias="taskName", description="Human-readable task name")
    status: TaskStatus = Field(
        default=TaskStatus.PENDING, description="Overall task status"
    )
    current_step: int = Field(
        0, ge=0, alias="currentStep", description="Index of current step"
    )
    total_steps: int = Field(
        0, ge=0, alias="totalSteps", description="Total number of steps"
    )
    steps: List[TaskStep] = Field(
        default_factory=list, description="List of task steps"
    )
    started_at: Optional[int] = Field(
        None, alias="startedAt", description="Task start timestamp (Unix ms)"
    )
    estimated_completion_ms: Optional[int] = Field(
        None, alias="estimatedCompletionMs", description="Estimated total duration in ms"
    )
    error: Optional[str] = Field(None, description="Error message if task failed")


# =============================================================================
# WIDGET STATE MODELS
# =============================================================================


class ProjectStatusState(BaseModel):
    """
    Project Status Widget State.

    Shows project progress with status indicator and task counts.
    """

    model_config = ConfigDict(populate_by_name=True, use_enum_values=True)

    project_id: str = Field(..., alias="projectId", description="Unique project identifier")
    name: str = Field(..., description="Project display name")
    status: ProjectStatus = Field(..., description="Current project status")
    progress: int = Field(..., ge=0, le=100, description="Progress percentage (0-100)")
    tasks_completed: int = Field(
        ..., ge=0, alias="tasksCompleted", description="Number of completed tasks"
    )
    tasks_total: int = Field(
        ..., ge=0, alias="tasksTotal", description="Total number of tasks"
    )
    last_updated: int = Field(
        ..., alias="lastUpdated", description="Last update timestamp (Unix ms)"
    )
    summary: Optional[str] = Field(None, description="Optional text summary")


class MetricEntry(BaseModel):
    """
    Single Metric Entry.

    Represents a single metric value with optional trend indicator.
    """

    model_config = ConfigDict(populate_by_name=True, use_enum_values=True)

    id: str = Field(..., description="Unique metric identifier")
    label: str = Field(..., description="Metric display label")
    value: Union[int, float, str] = Field(
        ..., description="Metric value (can be number or formatted string)"
    )
    unit: Optional[str] = Field(None, description="Optional unit suffix (e.g., '%', 'hrs')")
    trend: Optional[TrendDirection] = Field(None, description="Trend direction")
    change: Optional[str] = Field(None, description="Change description (e.g., '+5%')")
    change_percent: Optional[float] = Field(
        None, alias="changePercent", description="Change percentage value"
    )


class MetricsState(BaseModel):
    """
    Metrics Widget State.

    Container for multiple metrics with period context.
    """

    model_config = ConfigDict(populate_by_name=True)

    title: str = Field(default="Key Metrics", description="Widget title")
    metrics: List[MetricEntry] = Field(
        default_factory=list, description="Array of metric entries"
    )
    period: Optional[str] = Field(
        None, description="Time period description (e.g., 'Last 7 days')"
    )
    last_updated: int = Field(
        ..., alias="lastUpdated", description="Last update timestamp (Unix ms)"
    )


class ActivityEntry(BaseModel):
    """
    Activity Entry.

    Represents a single activity in the feed.
    """

    model_config = ConfigDict(populate_by_name=True)

    id: str = Field(..., description="Unique activity identifier")
    user: str = Field(..., description="User who performed the action")
    user_avatar: Optional[str] = Field(None, alias="userAvatar", description="User avatar URL")
    action: str = Field(..., description="Action description")
    target: Optional[str] = Field(
        None, description="Target of the action (e.g., task name)"
    )
    timestamp: int = Field(..., description="Activity timestamp (Unix ms)")
    project_id: Optional[str] = Field(None, alias="projectId", description="Related project ID")


class ActivityState(BaseModel):
    """
    Activity Widget State.

    Container for activity feed with pagination support.
    """

    model_config = ConfigDict(populate_by_name=True)

    activities: List[ActivityEntry] = Field(
        default_factory=list, description="Array of activity entries"
    )
    has_more: bool = Field(
        default=False, alias="hasMore", description="Whether more activities are available"
    )
    last_updated: int = Field(
        ..., alias="lastUpdated", description="Last update timestamp (Unix ms)"
    )


class AlertEntry(BaseModel):
    """
    Alert Entry.

    Represents a single alert notification.
    """

    model_config = ConfigDict(populate_by_name=True, use_enum_values=True)

    id: str = Field(..., description="Unique alert identifier")
    type: AlertType = Field(..., description="Alert severity type")
    title: str = Field(..., description="Alert title")
    message: str = Field(..., description="Alert message body")
    timestamp: int = Field(..., description="Alert timestamp (Unix ms)")
    dismissable: bool = Field(default=True, description="Whether the alert can be dismissed")
    dismissed: bool = Field(default=False, description="Whether the alert has been dismissed")
    action_label: Optional[str] = Field(
        None, alias="actionLabel", description="Optional action button label"
    )
    action_url: Optional[str] = Field(None, alias="actionUrl", description="Optional action URL")


# =============================================================================
# LOADING & ERROR STATE
# =============================================================================


class LoadingState(BaseModel):
    """
    Loading State.

    Tracks loading status and which agents are being queried.
    """

    model_config = ConfigDict(populate_by_name=True)

    is_loading: bool = Field(
        default=False, alias="isLoading", description="Whether any loading is in progress"
    )
    loading_agents: List[str] = Field(
        default_factory=list,
        alias="loadingAgents",
        description="List of agents currently being queried",
    )
    started_at: Optional[int] = Field(
        None, alias="startedAt", description="Loading start timestamp (Unix ms)"
    )


# =============================================================================
# ROOT DASHBOARD STATE
# =============================================================================


class WidgetsState(BaseModel):
    """
    Widget Container State.

    Holds all widget-specific state objects.
    """

    model_config = ConfigDict(populate_by_name=True)

    project_status: Optional[ProjectStatusState] = Field(
        None, alias="projectStatus", description="Project status widget data"
    )
    metrics: Optional[MetricsState] = Field(None, description="Metrics widget data")
    activity: Optional[ActivityState] = Field(None, description="Activity feed widget data")
    alerts: List[AlertEntry] = Field(
        default_factory=list, description="Alert entries (array, not single widget)"
    )


class DashboardState(BaseModel):
    """
    Root Dashboard State.

    This is the complete state object shared between agent and frontend
    via the AG-UI protocol's state synchronization mechanism.
    """

    model_config = ConfigDict(populate_by_name=True)

    version: int = Field(default=STATE_VERSION, description="Schema version for migrations")
    timestamp: int = Field(..., description="Last update timestamp (Unix ms)")
    active_project: Optional[str] = Field(
        None, alias="activeProject", description="Currently focused project ID"
    )
    workspace_id: Optional[str] = Field(
        None, alias="workspaceId", description="Current workspace ID"
    )
    user_id: Optional[str] = Field(None, alias="userId", description="Current user ID")

    # Widget data container
    widgets: WidgetsState = Field(
        default_factory=WidgetsState, description="Widget data container"
    )

    # Loading state
    loading: LoadingState = Field(
        default_factory=LoadingState, description="Loading state"
    )

    # Error state (agentId -> error message)
    errors: Dict[str, str] = Field(
        default_factory=dict, description="Error state (agentId -> error message)"
    )

    # Active tasks for progress tracking (DM-05.4)
    active_tasks: List[TaskProgress] = Field(
        default_factory=list,
        alias="activeTasks",
        description="Currently active long-running tasks",
    )

    @classmethod
    def create_initial(
        cls,
        workspace_id: Optional[str] = None,
        user_id: Optional[str] = None,
        active_project: Optional[str] = None,
    ) -> "DashboardState":
        """
        Create initial empty dashboard state.

        Used to initialize the state store and for tests.

        Args:
            workspace_id: Optional workspace/tenant ID
            user_id: Optional user ID
            active_project: Optional initially active project

        Returns:
            Fresh DashboardState with defaults
        """
        return cls(
            timestamp=int(datetime.utcnow().timestamp() * 1000),
            active_project=active_project,
            workspace_id=workspace_id,
            user_id=user_id,
        )

    def to_frontend_dict(self) -> Dict[str, Any]:
        """
        Convert to frontend-compatible dictionary with camelCase keys.

        This method serializes the state using the aliases defined on each field,
        ensuring the output matches the TypeScript schema exactly.

        Returns:
            Dictionary with camelCase keys suitable for frontend consumption
        """
        return self.model_dump(by_alias=True, exclude_none=True)

File: apps/web/src/lib/schemas/dashboard-state.ts
/**
 * Dashboard Shared State Schemas
 *
 * These schemas define the structure of state shared between
 * the Dashboard Gateway agent and the frontend via AG-UI protocol.
 * State is synchronized bidirectionally using CopilotKit's
 * useCoAgentStateRender mechanism.
 *
 * @see docs/modules/bm-dm/epics/epic-dm-04-tech-spec.md
 * Epic: DM-04 | Story: DM-04.1
 */
import { z } from 'zod';

// =============================================================================
// STATE VERSION
// =============================================================================

/**
 * State schema version for migrations.
 * Increment when making breaking changes to the schema structure.
 */
export const STATE_VERSION = 1;

// =============================================================================
// ENUMS
// =============================================================================

/**
 * Project status values
 */
export const ProjectStatusEnum = z.enum(['on-track', 'at-risk', 'behind', 'completed']);
export type ProjectStatusValue = z.infer<typeof ProjectStatusEnum>;

/**
 * Metric trend direction
 */
export const TrendDirectionEnum = z.enum(['up', 'down', 'neutral']);
export type TrendDirection = z.infer<typeof TrendDirectionEnum>;

/**
 * Alert severity type
 */
export const AlertTypeEnum = z.enum(['error', 'warning', 'info', 'success']);
export type AlertType = z.infer<typeof AlertTypeEnum>;

// =============================================================================
// TASK PROGRESS SCHEMAS (DM-05.4)
// =============================================================================

/**
 * Task step execution status
 */
export const TaskStepStatusEnum = z.enum(['pending', 'running', 'completed', 'failed']);
export type TaskStepStatus = z.infer<typeof TaskStepStatusEnum>;

/**
 * Overall task status
 */
export const TaskStatusEnum = z.enum(['pending', 'running', 'completed', 'failed', 'cancelled']);
export type TaskStatusValue = z.infer<typeof TaskStatusEnum>;

/**
 * Individual step within a task
 * Tracks execution state and progress of a single step in a multi-step task.
 */
export const TaskStepSchema = z.object({
  /** Step index (0-based) */
  index: z.number().int().min(0),
  /** Step display name */
  name: z.string(),
  /** Step execution status */
  status: TaskStepStatusEnum.default('pending'),
  /** Step start timestamp (Unix ms) */
  startedAt: z.number().optional(),
  /** Step completion timestamp (Unix ms) */
  completedAt: z.number().optional(),
  /** Sub-step progress percentage (0-100) */
  progress: z.number().min(0).max(100).optional(),
});

export type TaskStep = z.infer<typeof TaskStepSchema>;

/**
 * Progress state for a long-running task
 * Tracks overall task status and individual step progress for
 * real-time streaming from the agent.
 */
export const TaskProgressSchema = z.object({
  /** Unique task identifier */
  taskId: z.string(),
  /** Human-readable task name */
  taskName: z.string(),
  /** Overall task status */
  status: TaskStatusEnum.default('pending'),
  /** Index of current step */
  currentStep: z.number().int().min(0).default(0),
  /** Total number of steps */
  totalSteps: z.number().int().min(0).default(0),
  /** List of task steps */
  steps: z.array(TaskStepSchema).default([]),
  /** Task start timestamp (Unix ms) */
  startedAt: z.number().optional(),
  /** Task completion timestamp (Unix ms) */
  completedAt: z.number().optional(),
  /** Estimated total duration in ms */
  estimatedCompletionMs: z.number().optional(),
  /** Error message if task failed */
  error: z.string().optional(),
});

export type TaskProgress = z.infer<typeof TaskProgressSchema>;

// =============================================================================
// WIDGET STATE SCHEMAS
// =============================================================================

/**
 * Project Status Widget State
 * Shows project progress with status indicator and task counts.
 */
export const ProjectStatusStateSchema = z.object({
  /** Unique project identifier */
  projectId: z.string(),
  /** Project display name */
  name: z.string(),
  /** Current project status */
  status: ProjectStatusEnum,
  /** Progress percentage (0-100) */
  progress: z.number().min(0).max(100),
  /** Number of completed tasks */
  tasksCompleted: z.number().int().min(0),
  /** Total number of tasks */
  tasksTotal: z.number().int().min(0),
  /** Last update timestamp (Unix ms) */
  lastUpdated: z.number(),
  /** Optional text summary */
  summary: z.string().optional(),
});

export type ProjectStatusState = z.infer<typeof ProjectStatusStateSchema>;

/**
 * Single Metric Entry
 * Represents a single metric value with optional trend indicator.
 */
export const MetricEntrySchema = z.object({
  /** Unique metric identifier */
  id: z.string(),
  /** Metric display label */
  label: z.string(),
  /** Metric value (can be number or formatted string) */
  value: z.union([z.number(), z.string()]),
  /** Optional unit suffix (e.g., "%", "hrs") */
  unit: z.string().optional(),
  /** Trend direction */
  trend: TrendDirectionEnum.optional(),
  /** Change description (e.g., "+5%") */
  change: z.string().optional(),
  /** Change percentage value */
  changePercent: z.number().optional(),
});

export type MetricEntry = z.infer<typeof MetricEntrySchema>;

/**
 * Metrics Widget State
 * Container for multiple metrics with period context.
 */
export const MetricsStateSchema = z.object({
  /** Widget title */
  title: z.string().default('Key Metrics'),
  /** Array of metric entries */
  metrics: z.array(MetricEntrySchema),
  /** Time period description (e.g., "Last 7 days") */
  period: z.string().optional(),
  /** Last update timestamp (Unix ms) */
  lastUpdated: z.number(),
});

export type MetricsState = z.infer<typeof MetricsStateSchema>;

/**
 * Activity Entry
 * Represents a single activity in the feed.
 */
export const ActivityEntrySchema = z.object({
  /** Unique activity identifier */
  id: z.string(),
  /** User who performed the action */
  user: z.string(),
  /** User avatar URL */
  userAvatar: z.string().optional(),
  /** Action description */
  action: z.string(),
  /** Target of the action (e.g., task name) */
  target: z.string().optional(),
  /** Activity timestamp (Unix ms) */
  timestamp: z.number(),
  /** Related project ID */
  projectId: z.string().optional(),
});

export type ActivityEntry = z.infer<typeof ActivityEntrySchema>;

/**
 * Activity Widget State
 * Container for activity feed with pagination support.
 */
export const ActivityStateSchema = z.object({
  /** Array of activity entries */
  activities: z.array(ActivityEntrySchema),
  /** Whether more activities are available */
  hasMore: z.boolean().default(false),
  /** Last update timestamp (Unix ms) */
  lastUpdated: z.number(),
});

export type ActivityState = z.infer<typeof ActivityStateSchema>;

/**
 * Alert Entry
 * Represents a single alert notification.
 */
export const AlertEntrySchema = z.object({
  /** Unique alert identifier */
  id: z.string(),
  /** Alert severity type */
  type: AlertTypeEnum,
  /** Alert title */
  title: z.string(),
  /** Alert message body */
  message: z.string(),
  /** Alert timestamp (Unix ms) */
  timestamp: z.number(),
  /** Whether the alert can be dismissed */
  dismissable: z.boolean().default(true),
  /** Whether the alert has been dismissed */
  dismissed: z.boolean().default(false),
  /** Optional action button label */
  actionLabel: z.string().optional(),
  /** Optional action URL */
  actionUrl: z.string().optional(),
});

export type AlertEntry = z.infer<typeof AlertEntrySchema>;

// =============================================================================
// LOADING & ERROR STATE SCHEMAS
// =============================================================================

/**
 * Loading State
 * Tracks loading status and which agents are being queried.
 */
export const LoadingStateSchema = z.object({
  /** Whether any loading is in progress */
  isLoading: z.boolean().default(false),
  /** List of agents currently being queried */
  loadingAgents: z.array(z.string()).default([]),
  /** Loading start timestamp (Unix ms) */
  startedAt: z.number().optional(),
});

export type LoadingState = z.infer<typeof LoadingStateSchema>;

/**
 * Error State
 * Map of agent IDs to error messages.
 */
export const ErrorStateSchema = z.record(z.string(), z.string());

export type ErrorState = z.infer<typeof ErrorStateSchema>;

// =============================================================================
// ROOT DASHBOARD STATE SCHEMA
// =============================================================================

/**
 * Widget Container State
 * Holds all widget-specific state objects.
 */
export const WidgetsStateSchema = z.object({
  /** Project status widget data */
  projectStatus: ProjectStatusStateSchema.nullable().default(null),
  /** Metrics widget data */
  metrics: MetricsStateSchema.nullable().default(null),
  /** Activity feed widget data */
  activity: ActivityStateSchema.nullable().default(null),
  /** Alert entries (array, not single widget) */
  alerts: z.array(AlertEntrySchema).default([]),
});

export type WidgetsState = z.infer<typeof WidgetsStateSchema>;

/**
 * Root Dashboard State
 *
 * This is the complete state object shared between agent and frontend
 * via the AG-UI protocol's state synchronization mechanism.
 */
export const DashboardStateSchema = z.object({
  /** Schema version for migrations */
  version: z.number().default(STATE_VERSION),
  /** Last update timestamp (Unix ms) */
  timestamp: z.number(),
  /** Currently focused project ID */
  activeProject: z.string().nullable().default(null),
  /** Current workspace ID */
  workspaceId: z.string().optional(),
  /** Current user ID */
  userId: z.string().optional(),

  /** Widget data container */
  widgets: WidgetsStateSchema.default({
    projectStatus: null,
    metrics: null,
    activity: null,
    alerts: [],
  }),

  /** Loading state */
  loading: LoadingStateSchema.default({
    isLoading: false,
    loadingAgents: [],
  }),

  /** Error state (agentId -> error message) */
  errors: ErrorStateSchema.default({}),

  /** Currently active long-running tasks (DM-05.4) */
  activeTasks: z.array(TaskProgressSchema).default([]),
});

export type DashboardState = z.infer<typeof DashboardStateSchema>;

// =============================================================================
// STATE UPDATE SCHEMAS (Partial updates)
// =============================================================================

/**
 * Partial update for dashboard state.
 * Used for incremental state updates from agents.
 */
export const DashboardStateUpdateSchema = DashboardStateSchema.partial();

export type DashboardStateUpdate = z.infer<typeof DashboardStateUpdateSchema>;

// =============================================================================
// STATE VALIDATION HELPERS
// =============================================================================

/**
 * Validate dashboard state, returning typed result or null on failure.
 *
 * @param data - Unknown data to validate
 * @returns Validated DashboardState or null if invalid
 */
export function validateDashboardState(data: unknown): DashboardState | null {
  const result = DashboardStateSchema.safeParse(data);
  if (result.success) {
    return result.data;
  }
  console.error('[DashboardState] Validation failed:', result.error.issues);
  return null;
}

/**
 * Validate a partial dashboard state update.
 *
 * @param data - Unknown data to validate
 * @returns Validated DashboardStateUpdate or null if invalid
 */
export function validateDashboardStateUpdate(
  data: unknown
): DashboardStateUpdate | null {
  const result = DashboardStateUpdateSchema.safeParse(data);
  if (result.success) {
    return result.data;
  }
  console.error('[DashboardState] Update validation failed:', result.error.issues);
  return null;
}

/**
 * Create initial empty dashboard state.
 * Used to initialize the state store and for tests.
 *
 * @param options - Optional initial values
 * @returns Fresh DashboardState with defaults
 */
export function createInitialDashboardState(options?: {
  workspaceId?: string;
  userId?: string;
  activeProject?: string | null;
}): DashboardState {
  return {
    version: STATE_VERSION,
    timestamp: Date.now(),
    activeProject: options?.activeProject ?? null,
    workspaceId: options?.workspaceId,
    userId: options?.userId,
    widgets: {
      projectStatus: null,
      metrics: null,
      activity: null,
      alerts: [],
    },
    loading: {
      isLoading: false,
      loadingAgents: [],
    },
    errors: {},
    activeTasks: [],
  };
}
